ISISComputingGroup/ibex_gui
# Ticket2082 polaris jaws

### Description of work

This is a joint pull request for tickets #2082 and #2083

There are a lot of different bits of code here; sorry!

### To test

Review at the same time as these various other pull requests:
https://github.com/ISISComputingGroup/EPICS-motorExtensions/pull/1
https://github.com/ISISComputingGroup/EPICS-ioc/pull/129
https://github.com/ISISComputingGroup/EPICS-jaws/pull/3
https://github.com/ISISComputingGroup/EPICS-AccessSecurity/pull/1

The motors should be set up like this (in `jaws.cmd`):

```
############ Polaris jaws ##############
# Create a soft motor record on top of a normal motor record which allow the gap to be set for the motor
# Copy this file to C:\Instrument\Settings\config\<machine>\configurations\galil

# Define a set of jaws from first 4 motors on second controller (if present)
$(IFDMC02) dbLoadRecords("$(JAWS)/db/jaws.db","P=$(MYPVPREFIX)MOT:,JAWS=JAWS1:,mXN=MTR0201,mXS=MTR0202,mXW=MTR0203,mXE=MTR0204")
$(IFDMC02) dbLoadRecords("$(JAWS)/db/jaws.db","P=$(MYPVPREFIX)MOT:,JAWS=JAWS2:,mXN=MTR0205,mXS=MTR0206,mXW=MTR0207,mXE=MTR0208")
$(IFDMC03) dbLoadRecords("$(JAWS)/db/jaws.db","P=$(MYPVPREFIX)MOT:,JAWS=JAWS3:,mXN=MTR0301,mXS=MTR0302,mXW=MTR0303,mXE=MTR0304")
$(IFDMC03) dbLoadRecords("$(JAWS)/db/jaws.db","P=$(MYPVPREFIX)MOT:,JAWS=JAWS4:,mXN=MTR0305,mXS=MTR0306,mXW=MTR0307,mXE=MTR0308")
$(IFDMC04) dbLoadRecords("$(JAWS)/db/jaws.db","P=$(MYPVPREFIX)MOT:,JAWS=JAWS5:,mXN=MTR0401,mXS=MTR0402,mXW=MTR0403,mXE=MTR0404,ASG=MANAGER")

# Define jaws as soft motor records using the second galil controller, if present

$(IFDMC02) dbLoadRecords("C:/Instrument/Apps/EPICS/support/motorExtensions/master/db/polarisjawset.db", "P=$(MYPVPREFIX), COLHEIGHT=40, COLWIDTH=82, MOD_TO_SAMPLE=14000, MOD_TO_COL=1610, MOD_TO_JAWS1=6502, MOD_TO_JAWS2=9440, MOD_TO_JAWS3=11085, MOD_TO_JAWS4=11735, MOD_TO_JAWS5=13175")
```


### Acceptance criteria (one list for all the various pull requests)

- [ ] There is an OPI for controlling the Jaw set on POLARIS
- [ ] There are suitable DB records for controlling the Jaw set on POLARIS
- [ ] The DB records perform the required maths for the jaw set on polaris (see supporting documents on sharepoint site under Polaris/polarisCalculations. See also the calculations in the VI)
- [ ] The GUI has an option to go into "manager mode" (under the IBEX menu). ~~The password is initially set as aaa, when the reviewer is happy with everything I will update it to the final production password~~. The password is now the one that you would expect.
- [ ] The GUI displays the status of manager mode in the banner with suitable colours for enabled, disabled and disconnected.
- [ ] The password is stored as a salted hash using a hashing algorithm designed for passwords.
- [ ] Manager mode applies to the entire instrument, not just one client.
- [ ] If the instrument is in manager mode, you can control jaw set 5 on polaris (GUI, caput, genie_python).
- [ ] If the instrument is not in manager mode, you cannot control jaw set 5 on polaris (GUI, caput, genie_python).
- [ ] The access level of the jaw sets is controlled by a macro in `jaws.cmd`

---

#### Code Review


### Functional Tests


### Final Steps
